doc/sharding.md

```diff
@@ -0,0 +1,96 @@
+## M3DB Sharding and Replication
+
+### Sharding
+
+Data will be consistently hashed to a set of virtual nodes.  M3DB can be configured to use any hashing function and a configured number of shards.  For the sake of this documentation we’ll choose the following configuration:
+
+**Hash:** Murmur32
+
+**Shards:** 4096
+
+e.g. for a four node single replica set:
+
+Node 1: 0..1023
+
+Node 2: 1023..2047
+
+Node 3: 2048..3071
+
+Node 4: 3072..4095
+
+### Replication
+
+Full sets of data, that is data from all 0 - 4096 shards will be replicated on separate nodes that are rack aware.  That is, the set of datacenter racks that locate a replica’s data should be distinct to the racks that locate all other replica’s data.
+
+### Replica
+
+Each replica will have its own view of each shard:
+
+```
+Replica struct {
+  index uint32
+  shards []Shard
+}
+```
+
+### Shard state
+
+Each shard will have the following data:
+
+```
+Shard struct {
+  index uint32
+  assignments []ShardAssignment
+  minVersion uint64 
+}
+
+ShardAssignment struct {
+  host Host
+  state ShardState 
+}
+
+enum ShardState {
+  PROPOSED,
```

Comment:
I thought we agreed on having only two states `INITIALIZING` and `AVAILABLE`, do we still have three states?